### ISSUE TITLE
5% HF Regression from mem eff PR

### ISSUE THREAD

**Issue by eellison** [pytorch/pytorch#104356]
### 🐛 Describe the bug

There was a 5% performance regression from landing https://github.com/pytorch/pytorch/pull/103704. 

See comparison [here](https://hud.pytorch.org/benchmark/huggingface/inductor_with_cudagraphs?startTime=Wed,%2021%20Jun%202023%2016:53:23%20GMT&stopTime=Wed,%2028%20Jun%202023%2016:53:23%20GMT&granularity=hour&mode=training&dtype=amp&lBranch=main&lCommit=803c14490b189f9b755ecb9f2a969876088ea243&rBranch=main&rCommit=3c34a00d1b94aa4ceeede498ab878f1c5a25afb1). 

For example, BertForQuestionAnswering went from 2.3244 → 2.0836.

The root cause of this is discussed in the pr. Additional support was added for the memory efficient implementation with dropout, which led to us dispatching to it, which led to a 5% regression. We should 
a) improve heuristics
b) autotune which backend to use when MAX_AUTOTUNE is set

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @drisspg 

### Versions

master

**Comment by eellison** [comment]
This will be resolved by flash attention v2, which is universally faster.
